src/main/scala/io/pravega/connectors/spark/NonTransactionPravegaWriter.scala

```diff
@@ -1,220 +1,221 @@
+    NonTransactionPravegaWriterFactory(scopeName, streamName, clientConfig, schema)
+
+  override def commit(epochId: Long, messages: Array[WriterCommitMessage]): Unit = {}
+
+  override def abort(epochId: Long, messages: Array[WriterCommitMessage]): Unit = {}
+}
+
+/**
+  * A [[DataWriterFactory]] for Pravega writing. It will be serialized and sent to executors to
+  * generate the per-task data writers.
+  *
+  * @param schema                   The schema of the input data.
+  */
+case class NonTransactionPravegaWriterFactory(
+                                       scopeName: String,
+                                       streamName: String,
+                                       clientConfig: ClientConfig,
+                                       schema: StructType)
+  extends DataWriterFactory with StreamingDataWriterFactory with Logging {
+
+  override def createWriter(partitionId: Int, taskId: Long): DataWriter[InternalRow] = {
+    new NonTransactionPravegaDataWriter(scopeName, streamName, clientConfig, schema)
+  }
+
+  override def createWriter(partitionId: Int, taskId: Long, epochId: Long): DataWriter[InternalRow] = {
+    new NonTransactionPravegaDataWriter(scopeName, streamName, clientConfig, schema)
+  }
+}
+
+/**
+  * A [[DataWriter]] for Pravega writing. One data writer will be created in each partition to
+  * process incoming rows.
+  *
+  * @param inputSchema              The attributes in the input data.
+  */
+class NonTransactionPravegaDataWriter(
+                               scopeName: String,
+                               streamName: String,
+                               clientConfig: ClientConfig,
+                               inputSchema: StructType)
+  extends DataWriter[InternalRow] with Logging {
+
+  val pendingWriteCounts = new AtomicInteger(0)
+
+  private val projection = createProjection
+
+  private val clientFactory = EventStreamClientFactory.withScope(scopeName, clientConfig)
+  private val writer: EventStreamWriter[ByteBuffer] = clientFactory.createEventWriter(
+    streamName,
+    new ByteBufferSerializer,
+    EventWriterConfig
+      .builder
+      .build)
+
+
+  override def write(row: InternalRow): Unit = {
+    val projectedRow = projection(row)
+    val event = projectedRow.getBinary(1)
+    val eventToLog = if (log.isDebugEnabled) {
+      val s = new String(event, StandardCharsets.UTF_8)
+      val maxLength = 100
+      if (s.length > maxLength) s.substring(0, maxLength) + "..." else s
+    } else null
+
+    log.info(s"write: begin")
+
+    val haveRoutingKey = !projectedRow.isNullAt(0)
+    if (haveRoutingKey) {
+      val routingKey = projectedRow.getUTF8String(0).toString
+      log.debug(s"write: routingKey=${routingKey}, event=${eventToLog}")
+      writer.writeEvent(routingKey, ByteBuffer.wrap(event))
+    } else {
+      log.debug(s"write: event=${eventToLog}")
+
+      val f = toScala(writer.writeEvent(ByteBuffer.wrap(event)))
+      f

```

Comment:
We also need to log any errors from the haveRoutingKey=true case. Do something like this:
```
final CompletableFuture writeFuture;
if (haveRoutingKey) {
   ...
   writeFuture = writer.writeEvent(routingKey, ByteBuffer.wrap(event));
} else {
   ...
   writeFuture = writer.writeEvent(ByteBuffer.wrap(event));
}
// Log any errors immediately but do not handle them here. If any errors occur, commit() will eventually return an error.
toScala(writeFuture).recover { case e =>
          log.error(s"write: event=${eventToLog}, error=${e}")
        }
```